File: applications/luci-app-statistics/luasrc/model/cbi/luci_statistics/df.lua
--[[

Luci configuration model for statistics - collectd df plugin configuration
(c) 2008 Freifunk Leipzig / Jo-Philipp Wich <xm@leipzig.freifunk.net>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

$Id$

]]--

m = Map("luci_statistics",
	translate("DF Plugin Configuration"),
	translate(
		"The df plugin collects statistics about the disk space " ..
		"usage on different devices, mount points or filesystem types."
	))

-- collectd_df config section
s = m:section( NamedSection, "collectd_df", "luci_statistics" )

-- collectd_df.enable
enable = s:option( Flag, "enable", translate("Enable this plugin") )
enable.default = 0

-- collectd_df.devices (Device)
devices = s:option( Value, "Devices", translate("Monitor devices") )
devices.default  = "/dev/mtdblock/4"
devices.optional = true
devices:depends( "enable", 1 )

-- collectd_df.mountpoints (MountPoint)
mountpoints = s:option( Value, "MountPoints", translate("Monitor mount points") )
mountpoints.default  = "/overlay"
mountpoints.optional = true
mountpoints:depends( "enable", 1 )

-- collectd_df.fstypes (FSType)
fstypes = s:option( Value, "FSTypes", translate("Monitor filesystem types") )
fstypes.default  = "tmpfs"
fstypes.optional = true
fstypes:depends( "enable", 1 )

-- collectd_df.ignoreselected (IgnoreSelected)
ignoreselected = s:option( Flag, "IgnoreSelected", translate("Monitor all except specified") )
ignoreselected.default = 0
ignoreselected:depends( "enable", 1 )

return m

File: applications/luci-app-statistics/htdocs/luci-static/resources/view/statistics/plugins/df.js
'use strict';
'require baseclass';
'require fs';
'require form';

return baseclass.extend({
	title: _('DF Plugin Configuration'),
	description: _('The df plugin collects statistics about the disk space usage on different devices, mount points or filesystem types.'),

	addFormOptions(s) {
		let o;

		o = s.option(form.Flag, 'enable', _('Enable this plugin'));

		o = s.option(form.DynamicList, 'Devices', _('Monitor devices'));
		o.optional = true;
		o.depends('enable', '1');
		o.load = function(section_id) {
			return fs.lines('/proc/partitions').then(L.bind(function(lines) {
				const parts = [];

				for (let l of lines) {
					const line = L.toArray(l);
					if (!isNaN(line[0]))
						parts.push('/dev/' + line[3]);
				}

				parts.sort();

				for (let p of parts)
					this.value(p);

				return this.super('load', [section_id]);
			}, this));
		};

		o = s.option(form.DynamicList, 'MountPoints', _('Monitor mount points'));
		o.default = '/overlay';
		o.optional = true;
		o.depends('enable', '1');
		o.load = function(section_id) {
			return fs.lines('/proc/mounts').then(L.bind(function(lines) {
				let mounts = {};

				for (let l of lines) {
					const line = L.toArray(l);
					mounts[line[1]] = true;
				}

				mounts = Object.keys(mounts).sort();

				for (let m of mounts)
					this.value(m);

				return this.super('load', [section_id]);
			}, this));
		};

		o = s.option(form.DynamicList, 'FSTypes', _('Monitor filesystem types'));
		o.default = 'tmpfs';
		o.optional = true;
		o.depends('enable', '1');
		o.load = function(section_id) {
			return Promise.all([
				fs.lines('/etc/filesystems'),
				fs.lines('/proc/filesystems')
			]).then(L.bind(function(lines) {
				const fslines = lines[0].concat(lines[1]);
				let fstypes = {};

				for (let fsl of fslines) {
					var line = L.toArray(fsl);

					if (line.length == 2 && line[0] == 'nodev')
						continue;

					fstypes[line.pop()] = true;
				}

				fstypes = Object.keys(fstypes).sort();

				for (let fst of fstypes)
					this.value(fst);

				return this.super('load', [section_id]);
			}, this));
		};

		o = s.option(form.Flag, 'IgnoreSelected', _('Monitor all except specified'));
		o.depends('enable', '1');

		o = s.option(form.Flag, 'ValuesPercentage', _('Free space, reserved space and used space is reported as relative values'));
		o.depends('enable', '1');
	},

	configSummary(section) {
		const devs = L.toArray(section.Devices);
		const mounts = L.toArray(section.MountPoints);
		const fstypes = L.toArray(section.FSTypes);
		const count = devs.length + mounts.length + fstypes.length;
		const invert = section.IgnoreSelected == '1';

		if (count == 0)
			return _('Monitoring all partitions');
		else
			return (invert ? _('Monitoring all except %s, %s, %s') : _('Monitoring %s, %s, %s')).format(
				N_(devs.length, 'one device', '%d devices').format(devs.length),
				N_(mounts.length, 'one mount', '%d mounts').format(mounts.length),
				N_(fstypes.length, 'one filesystem type', '%d filesystem types').format(fstypes.length)
			);
	}
});
